client/network/src/gossip.rs

```diff
@@ -0,0 +1,312 @@
+	{
+		self.lock_queue().await.push_or_discard(message);
+	}
+}
+
+impl<M> fmt::Debug for DirectedGossip<M> {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_struct("DirectedGossip").finish()
+	}
+}
+
+impl<M> Drop for DirectedGossip<M> {
+	fn drop(&mut self) {
+		// The "clean" way to notify the `Condvar` here is normally to first lock the `Mutex`,
+		// then notify the `Condvar` while the `Mutex` is locked. Unfortunately, the `Mutex`
+		// being asynchronous, it can't reasonably be locked from within a destructor.
+		// For this reason, this destructor is a "best effort" destructor.
+		// See also the corresponding code in the background task.
+		self.shared.stop_task.store(true, atomic::Ordering::Release);
+		self.shared.condvar.notify_all();
+	}
+}
+
+/// Utility. Generic over the type of the messages. Holds a [`NetworkService`] and a [`PeerId`].
+/// Provides a [`DirectedGossipPrototype::build`] function that builds a [`DirectedGossip`].
+#[derive(Clone)]
+pub struct DirectedGossipPrototype {
+	service: Arc<dyn AbstractNotificationSender + Send + Sync + 'static>,
+	peer_id: PeerId,
+}
+
+impl DirectedGossipPrototype {
+	/// Builds a new [`DirectedGossipPrototype`] containing the given components.
+	pub fn new<B, H>(
+		service: Arc<NetworkService<B, H>>,
+		peer_id: PeerId,
+	) -> Self
+	where
+		B: BlockT + 'static,
+		H: ExHashT,
+	{
+		DirectedGossipPrototype {
+			service,
+			peer_id,
+		}
+	}
+
+	/// Turns this [`DirectedGossipPrototype`] into a [`DirectedGossip`] and a future.
+	///
+	/// See [`DirectGossip::new`] for details.
+	pub fn build<M, F>(
+		self,
+		protocol: ConsensusEngineId,
+		queue_size_limit: usize,
+		messages_encode: F
+	) -> (DirectedGossip<M>, impl Future<Output = ()> + Send + 'static)
+	where
+		M: Send + 'static,
+		F: Fn(M) -> Vec<u8> + Send + 'static,
+	{
+		let shared = Arc::new(Shared {
+			stop_task: atomic::AtomicBool::new(false),
+			condvar: Condvar::new(),
+			queue_size_limit,
+			messages_queue: Mutex::new(VecDeque::with_capacity(queue_size_limit)),
+		});
+
+		let task = spawn_task(
+			self.service,
+			self.peer_id,
+			protocol,
+			shared.clone(),
+			messages_encode
+		);
+
+		(DirectedGossip { shared }, task)
+	}
+}
+
+impl fmt::Debug for DirectedGossipPrototype {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_struct("DirectedGossipPrototype")
+			.field("peer_id", &self.peer_id)
+			.finish()
+	}
+}
+
+/// Locked queue of messages to the given peer.
+///
+/// As long as this struct exists, the background task is asleep and the owner of the [`QueueLock`]
+/// is in total control of the buffer.
+#[must_use]
+pub struct QueueLock<'a, M> {
+	messages_queue: MutexGuard<'a, VecDeque<M>>,
+	condvar: &'a Condvar,
+	/// Same as [`Shared::queue_size_limit`].
+	queue_size_limit: usize,
+}
+
+impl<'a, M: Send + 'static> QueueLock<'a, M> {
+	/// Pushes a message to the queue, or discards it if the queue is full.
+	pub fn push_or_discard(&mut self, message: M) {
```

Comment:
Same here. It would be good to get back the `message` if the queue is full

Reply:
But what would you do with the message that is returned? Put it in another queue?
The only sane thing you can do when the queue is full is either discard the message or force-close the connection.

Additionally, what if the connection with the remote is closed? Are we supposed to return back the message as well? If so, then it's very problematic because we can't detect this.

Reply:
> The only sane things you can do when the queue is full is either discard the message or force-close the connection.

Isn't `retain` another option? We could decide to do that based on the type of the message we are trying to send.

And is there no sane thing you can do? Why can't you wait for space to appear in the queue or something like that?

Reply:
The way I see `retain` is that you'd call it all the time, even if there is space in the buffer. As far as I can tell, when a message is obsolete, there is no point in leaving it in the queue anyway.

> Why can't you wait for space to appear in the queue or something like that?

The entire reason for this API to exist is to remove the need for any waiting. See also [this paragraph](https://github.com/paritytech/substrate/blob/270a992bf299cd41b2ed4e211814d2a958e3bfc8/client/network/src/service.rs#L653-L671).

Ultimately there *has to* be a code somewhere that holds some sort of `HashMap<PeerId, DirectedGossip>`. If it needs to send a message to one of the peers and its buffer is full, then it shouldn't wait for this peer and instead continue its processing.

Reply:
It would be nice to have a way to wait, although I agree that the API for that should be used sparingly so you don't degrade to the performance of the slowest peer. There are cases where we don't want to drop messages, for instance when responding to a validator's request.

Reply:
> There are cases where we don't want to drop messages

I believe that everything that would fall in this category should be covered by request-response protocols.

Rather than adding a wait, I could restore [the `push_unbounded` method that I've removed after a review](https://github.com/paritytech/substrate/pull/6803/commits/c11bcfb5cecd983085eacf3f2b08f9a6e27c6651#diff-f5e0015f0355ca80bb672b3980b8eba1L218-L223).

Reply:
@rphmeier @tomaka is this conversation resolved?